### DISCUSSION TITLE
Protocol handler to join network games

### DISCUSSION THREAD

**Discussion by nielsmh** [OpenTTD/OpenTTD#11741] [Features]
It could be some good UI to support a protocol handler to join network games.
A protocol handler could allow players to make links to a server directly from a website or social service and have players join through that. It would also make the browser at http://servers.openttd.org/ more accessible.

## URI scheme
Example of what an OpenTTD server link could look like:
`openttd://127.0.0.1:3981?version=1.9.3&password=asdf&join=4&companypassword=qwerty`

Suggested parts:
- `openttd:` because a scheme name is required
- `//` because everyone else does it
- `127.0.0.1:3981` IP address (or hostname) and port
- `?` make a HTTP-like query string
- `version=1.9.3` could in the future allow having a meta-handler that selects between multiple installed versions of the game for the required one
- `password=` server password, leave out if not required (prompt user if server needs one or the provided is wrong)
- `join=` join a company (leave out to spectate)
- `join=new` create a new company on connect, if possible
- `companypassword=` password to join company, or to set on the new company created (spectate if provided password is wrong)

## Implementation
Every OS has its own way of registering URI handlers, the installers/packaging would have to handle that.

The implementation in game would be a commandline parameter such as `-U <uri>`. The game should perhaps ask the player for confirmation before beginning to connect, so they don't accidentally connect to random servers.

**Comment by sdcloudt** [comment]
Working on it.

**Comment by JMcKiern** [comment]
I was looking into this over the weekend. I have a working parser but I couldn't figure out how to call [`NetworkClientConnectGame()`](https://github.com/OpenTTD/OpenTTD/blob/faaec445d669a222718c2634eb059c44c00a43b7/src/network/network.cpp#L664) at the right time. I tried calling it before [`VideoDriver::GetInstance()->MainLoop()`](https://github.com/OpenTTD/OpenTTD/blob/a38122e6f21374440f5d644b5d54b94ebfac4300/src/openttd.cpp#L858) to see what would happen but it didn't seem to work. Maybe there is some network initialization that needs to be performed beforehand?

Presumably, the main menu needs to load first and *then* this function needs to be called. 
Any ideas on how to "queue" this join game command?

Is there perhaps a function similar to [`ShowQuery()`](https://github.com/OpenTTD/OpenTTD/blob/13cc8a0ceec90def39cbcb84135a0bf039793a6f/src/misc_gui.cpp#L1260) that would allow passing of the game server info to the callback function?

**Comment by nielsmh** [comment]
Have you looked at how the "traditional" command line parameters for joining a network game from the command line are handled in `openttd_main()` in openttd.cpp?

**Comment by JMcKiern** [comment]
Oh yeah, it's the AfterNewGRFScan variable, right? Don't know why I didn't look into that... I'll give it another go when I get some time
